### PR TITLE
Integrate clad

### PR DESCRIPTION
**Pull request by vgvassilev**
clad is a C++ plugin for clang that implements automatic differentiation of user-defined functions by employing the chain rule in forward mode, coupled with source code transformation and AST constant fold.

### CODE REVIEW

File: interpreter/cling/test/CodeGeneration/Symbols.C [hunk 1]
```diff
@@ -7,7 +7,7 @@
 //------------------------------------------------------------------------------
 
 // RUN: clang -shared %fPIC -DCLING_EXPORT=%dllexport -DBUILD_SHARED %s -o%T/libSymbols%shlibext
-// RUN: %cling --nologo -L%T -lSymbols  %s | FileCheck %s
+// RUN: %cling --nologo -L%T "Symbols"  %s | FileCheck %s
```
Comment (Axel-Naumann) [inline interpreter/cling/test/CodeGeneration/Symbols.C]: Does this need to be quoted?

File: interpreter/cling/lib/Interpreter/InvocationOptions.cpp [hunk 2]
```diff
@@ -92,7 +92,6 @@ static const char kNoStdInc[] = "-nostdinc";
 
   static void ParseLinkerOpts(cling::InvocationOptions& Opts,
                               InputArgList& Args /* , Diags */) {
-    Extend(Opts.LibsToLoad, Args.getAllArgValues(OPT_l));
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/InvocationOptions.cpp]: This change reasons that `-l` is less useful than `-load`. I would argue that
- `-l` is much more user-centric and much more common,
- `-L` without `-l` is surprising.

I'd be in favor or using `--load-plugin` or similar.

File: interpreter/cling/tools/CMakeLists.txt [hunk 3]
```diff
@@ -6,11 +6,14 @@
...
-
 if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/../lib/UserInterface/textinput
    OR CLING_INCLUDE_TESTS)
   add_subdirectory(driver)
   add_subdirectory(Jupyter)
   add_subdirectory(libcling)
   add_subdirectory(demo)
 endif()
+
+if(CLING_BUILD_PLUGINS)
```
Comment (Axel-Naumann) [inline interpreter/cling/tools/CMakeLists.txt]: Do we really need to make this conditional?

File: interpreter/cling/tools/plugins/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#
+#LISTSUBDIRS(subdirs ${CMAKE_CURRENT_SOURCE_DIR}/plugins/)
+#foreach(subdir ${SUBDIRS})
+#  add_subdirectory(${subdir})
+#endforeach()
+
+ExternalProject_Add(
+  clad
+  GIT_REPOSITORY https://github.com/vgvassilev/clad.git
+  GIT_TAG master
```
Comment (Axel-Naumann) [inline interpreter/cling/tools/plugins/CMakeLists.txt]: I don't think this can fly. ROOT 6.14.00 compiled in 2020 will still check out the master of clad, and that will fail to build against 2018's cling, llvm and clang. Maybe use a commit sha1 here?

File: interpreter/cling/tools/plugins/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+#  endforeach()
+#  set(${result} ${dirlist} PARENT_SCOPE)
+#endfunction()
+#
+#LISTSUBDIRS(subdirs ${CMAKE_CURRENT_SOURCE_DIR}/plugins/)
+#foreach(subdir ${SUBDIRS})
+#  add_subdirectory(${subdir})
+#endforeach()
+
+ExternalProject_Add(
```
Comment (Axel-Naumann) [inline interpreter/cling/tools/plugins/CMakeLists.txt]: Shouldn't this be in `clad/CMakeLists,txt`?

File: interpreter/cling/lib/Interpreter/CIFactory.cpp [hunk 6]
```diff
@@ -657,8 +657,17 @@ static void stringifyPreprocSetting(PreprocessorOptions& PPOpts,
       P.resize(ExeIncl.size());
       // Get foo/include
       llvm::sys::path::append(P, "include");
-      if (llvm::sys::fs::is_directory(P.str()))
+      if (llvm::sys::fs::is_directory(P.str())) {
         utils::AddIncludePaths(P.str(), HOpts, nullptr);
+        llvm::sys::path::append(P, "clang");
+        if (!llvm::sys::fs::is_directory(P.str())) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: What does this have to do with clad? Can this be moved into a separate PR?

Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/CIFactory.cpp]: This issue was discovered by running `CLING=.../cling ... -Xclang -load clad.so make check-cling`. IMO it is (remotely?) related to the current work.

File: interpreter/cling/test/CodeGeneration/Symbols.C (hunk 1 above)
Comment (vgvassilev) [reply interpreter/cling/test/CodeGeneration/Symbols.C]: No, it can go either way.

File: interpreter/cling/tools/plugins/CMakeLists.txt (hunk 4 above)
Comment (vgvassilev) [reply interpreter/cling/tools/plugins/CMakeLists.txt]: Sure, that's not there to stay for eternity. This is more like work in progress. I am planning to tag once we have a tiny demonstrator (say in ROOT math).

File: interpreter/cling/tools/CMakeLists.txt (hunk 3 above)
Comment (vgvassilev) [reply interpreter/cling/tools/CMakeLists.txt]: The intend was to not force cling standalone users to use this as it is quite likely that they won't need it.

File: interpreter/CMakeLists.txt [hunk 7]
```diff
@@ -9,6 +9,22 @@ set(LLVM_BUILD_TOOLS OFF CACHE BOOL "")
...
+  set(prefix ${CMAKE_INSTALL_PREFIX})
+else()
+  set(prefix ${ROOTSYS})
+endif()
+if(IS_ABSOLUTE ${CMAKE_INSTALL_SYSCONFDIR})
+  set(etcdir ${CMAKE_INSTALL_SYSCONFDIR})
+else()
+  set(etcdir ${prefix}/${CMAKE_INSTALL_SYSCONFDIR})
+endif()
+set(CLING_PLUGIN_INSTALL_PREFIX "${etcdir}/cling/plugins/" CACHE STRING "" FORCE)
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: `<prefix>/etc` is the wrong place for plugins. Only small text configuration files should go into `<prefix>/etc`.
Better choices would be `<prefix>/lib/cling` or `<prefix>/share/cling/plugins`. ROOT would make my `/etc` about 10 times bigger, for instance:
```
epsft-53 ~ # du -shc /etc
15M	/etc
15M	total
epsft-53 ~ # du -shc /opt/root/6.12/etc
116M	/opt/root/6.12/etc
116M	total
```

File: interpreter/cling/tools/plugins/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,51 @@
+#------------------------------------------------------------------------------
+# CLING - the C++ LLVM-based InterpreterG :)
```
Comment (amadio) [inline interpreter/cling/tools/plugins/CMakeLists.txt]: There is a typo above, and I belive the smiley is a bit out of place.

File: interpreter/cling/tools/plugins/CMakeLists.txt (hunk 5 above)
Comment (amadio) [reply interpreter/cling/tools/plugins/CMakeLists.txt]: Does this mean that an internet connection is always required to build ROOT when `clad` is enabled?

File: interpreter/cling/tools/plugins/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  clad
+  GIT_REPOSITORY https://github.com/vgvassilev/clad.git
+  GIT_TAG master
+  UPDATE_COMMAND ""
+  # TIMEOUT 10
+  # # Force separate output paths for debug and release builds to allow easy
+  # # identification of correct lib in subsequent TARGET_LINK_LIBRARIES commands
+  # CMAKE_ARGS -DCMAKE_ARCHIVE_OUTPUT_DIRECTORY_DEBUG:PATH=DebugLibs
+  #            -DCMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELEASE:PATH=ReleaseLibs
+  #            -Dgtest_force_shared_crt=ON
```
Comment (amadio) [inline interpreter/cling/tools/plugins/CMakeLists.txt]: Why all these comments above?

File: interpreter/cling/tools/plugins/CMakeLists.txt (hunk 5 above)
Comment (vgvassilev) [reply interpreter/cling/tools/plugins/CMakeLists.txt]: Yes, the usual story as the other cases of `ExternalProject_Add`.

File: interpreter/CMakeLists.txt (hunk 7 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: IIUC, our ROOT-based plugins live there, too.

File: interpreter/cling/tools/plugins/CMakeLists.txt (hunk 8 above)
Comment (vgvassilev) [reply interpreter/cling/tools/plugins/CMakeLists.txt]: Sorry, I cannot see the typo. Could you be more specific. About the smiley it is courtesy of @Axel-Naumann.

File: interpreter/cling/lib/Interpreter/InvocationOptions.cpp (hunk 2 above)
Comment (vgvassilev) [reply interpreter/cling/lib/Interpreter/InvocationOptions.cpp]: In theory, I can drop this commit and hope that we do not do anything more complicated which requires `-load` or any `-l` clang option.

File: interpreter/cling/tools/plugins/CMakeLists.txt (hunk 5 above)
Comment (amadio) [reply interpreter/cling/tools/plugins/CMakeLists.txt]: Not really, as the other external projects can also be picked up from the system instead of being downloaded. I did not see such an option for clad. In any case, we already have many requests for making ROOT build without requiring an internet connection, and we even have [ROOT-9163](https://sft.its.cern.ch/jira/browse/ROOT-9163) to track progress towards it, and this goes in the opposite direction. That is, I'd rather not make it harder and harder to fix ROOT-9163 if I could.

File: interpreter/cling/tools/plugins/CMakeLists.txt (hunk 8 above)
Comment (amadio) [reply interpreter/cling/tools/plugins/CMakeLists.txt]: `InterpreterG`

File: interpreter/CMakeLists.txt (hunk 7 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: True, but they will also need to be moved to a more appropriate place. The situation above is not acceptable. We need to start minimally following the FHS from now on, to be able to reach the point in which we will become a regular package in most Linux distributions. This is another point for “let's not make it worse if we can avoid it”.

Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: Ok, I'd pick `<prefix>/share/cling/plugins`.

Comment (amadio) [reply interpreter/CMakeLists.txt]: Hi Vassil, thanks for understanding. How many such plugins exist now, and are users expected to install more plugins separately? Actually, I have at least these on my machine:
```
/usr/lib64/krb5/plugins
/usr/lib64/qt4/plugins
/usr/lib64/nsbrowser/plugins
/usr/lib64/libv4l/plugins
/usr/lib64/qt5/plugins
/usr/lib64/bluetooth/plugins
/usr/lib64/peas-demo/plugins
/usr/lib64/eog/plugins
/usr/lib64/traceevent/plugins
```
Looking at [FHS](http://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html) to check, `<prefix>/share` is meant for architecture-independent data files, so I'm inclined to use `<prefix>/lib/root/plugins` for ROOT, and suggest that we use `<prefix>/lib/cling/plugins` for cling for consistency, what do you think? The rationale is that these are executables, so probably `<prefix>/lib` makes more sense.

Comment (pcanal) [reply interpreter/CMakeLists.txt]: Hi Guilherme,

I understand and agree with the principle of 'not making things harder for an upcoming change'.  In this case, however, doesn't putting Clad's file in a 'different' (but better) place than the others more confusing than helpful.  (i.e. it will help the developer later but will confused the users now).   What do you think and/or what am I missing?

Cheers,
Philippe.

Comment (amadio) [reply interpreter/CMakeLists.txt]: Hi Philippe, I disagree that installing files where FHS indicates is confusing. There is a problem that /etc is hard-coded in a few places for the other files now, so if this goes to the right place since the beginning it will be easier to fix the rest later, and it will avoid the appearance of more hard-coded paths where it only works with the default value. For example, changing SYSCONFDIR breaks ROOT right now, that's why I couldn't change it yet.

Comment (pcanal) [reply interpreter/CMakeLists.txt]: > I disagree that installing files where FHS indicates is confusing. 

You might be misunderstanding my point.  I agree with you that "where FHS indicates" is the better place to install the files and what we should strive for.  My concern is the in-between situation where/when we have some files installed "where FHS indicates" and some in other places.  (i.e semantically similar things installed in two different places) ... if we **already** have the case that some files are installed "where FHS indicates" then my point is  moot (because the confusing situation I fear is already there).

**Comment by vgvassilev** [comment]
I should have addressed some of the comments here and incorporated the feedback offline.

**Comment by vgvassilev** [comment]
@phsft-bot build!
